File: src/renderer/src/api/youtube.ts
import axios from 'axios'
import { getBaseURL } from './client'

function getClient() {
  return axios.create({ baseURL: `${getBaseURL()}/api`, timeout: 30000 })
}

export interface VideoInfo {
  title: string
  duration: number
  channel: string
}

export async function fetchVideoInfo(url: string): Promise<VideoInfo> {
  const res = await getClient().post('/youtube/info', { url })
  return res.data
}

export async function extractAudio(
  url: string,
  info: VideoInfo,
  startSec?: number,
  endSec?: number,
  signal?: AbortSignal
): Promise<{ audioBlob: Blob; title: string; duration: number }> {
  const res = await getClient().post(
    '/youtube/extract-audio',
    {
      url,
      title: info.title,
      duration: info.duration,
      start_sec: startSec ?? null,
      end_sec: endSec ?? null,
    },
    {
      responseType: 'blob',
      timeout: 120000,
      signal,
    }
  )

  const title = decodeURIComponent(res.headers['x-video-title'] || info.title)
  const duration = parseInt(res.headers['x-video-duration'] || '0', 10) || info.duration

  return {
    audioBlob: new Blob([res.data], { type: 'audio/wav' }),
    title,
    duration,
  }
}

File: python/app/routers/youtube.py
from urllib.parse import quote

from fastapi import APIRouter, HTTPException
from fastapi.responses import FileResponse
from pydantic import BaseModel
from starlette.background import BackgroundTask

from app.services.youtube_service import (
    cleanup_temp_file,
    extract_audio,
    get_video_info,
)

router = APIRouter(prefix="/youtube", tags=["youtube"])


class YouTubeURLRequest(BaseModel):
    url: str


class VideoInfoResponse(BaseModel):
    title: str
    duration: int
    channel: str


@router.post("/info", response_model=VideoInfoResponse)
async def fetch_video_info(req: YouTubeURLRequest):
    try:
        info = await get_video_info(req.url)
        return info
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


class ExtractAudioWithInfoRequest(BaseModel):
    url: str
    title: str = "Unknown"
    duration: int = 0
    start_sec: float | None = None
    end_sec: float | None = None


@router.post("/extract-audio")
async def extract_video_audio(req: ExtractAudioWithInfoRequest):
    try:
        audio_path = await extract_audio(req.url, req.start_sec, req.end_sec)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    return FileResponse(
        path=str(audio_path),
        media_type="audio/wav",
        filename="youtube_audio.wav",
        headers={
            "X-Video-Title": quote(req.title),
            "X-Video-Duration": str(req.duration),
        },
        background=BackgroundTask(cleanup_temp_file, audio_path),
    )
